### DISCUSSION TITLE
Can it handle org-roam-v2 ID's?

### DISCUSSION THREAD

**Discussion by buhtz** [kaushalmodi/ox-hugo#581] [Q&A] [answered]
Do I understand the documentation and the great examples right, that ox-hugo is able to handle the ID-links created by org-roam version2 correct?

Because ox-publish (org-mode in-build) is currently not able to handle that correct; sometimes yes, sometimes no. It is a known bug.

Let me give you an example of an org file with two links in it
```
:PROPERTIES:
:ID:       cb53b5cc-31b8-40ad-87c4-4cf3c6e3cec6
:END:
#+title: foobar
+date: [2022-03-09 Mi 10:20]
Two links to foo and bar
Link to foo: [[id:13ec1e89-abc9-4d43-a4a4-3005a1c9dfc7][foo]]
Link to bar: [[id:0c66c239-5b88-40bb-a0f1-0e15d8477c47][bar]]

```
But `org-hugo-export-wim-to-md` generates this where can see at the end that there are no links.```

```
+++
title = "foobar"
draft = false
+++

+date: <span class="timestamp-wrapper"><span class="timestamp">[2022-03-09 Mi 10:20]</span></span> <br/>
Two links to foo and bar <br/>
Link to foo:  <br/>
Link to bar:  <br/>
```

**Comment by kaushalmodi** [comment]
Can you provide a complete example with all the Org files that have those IDs assigned? If the ID targets are not found, you should be seeing an error by default. So I am surprised that it exported with no links.

Once you share a complete minimal reproducible example, we can debug.

Also, look at this example of linking Org documents using IDs: https://github.com/kaushalmodi/ox-hugo/tree/main/test/site/content-org/org-roam. That example is part of the `ox-hugo` test suite and it is passing.

**Reply by buhtz** [reply]
Thanks for your feedback. Good to hear that it should be possible. ;)
Before providing a good example I have to learn some ox-hugo basics. See my other questions.

**Comment by buhtz** [comment]
[example.zip](https://github.com/kaushalmodi/ox-hugo/files/8243668/example.zip)

Here you see an example with three org-roam-v2 nodes as org-, md- and html-file. As you see in the org-files "Foo" links to "Bar" and "Foobar" links to "Foo" and "Bar".

This happens from time to time with example projects (only this three files) sometimes in my original org-roam notes. Sometimes mixed where some links are present and some not.

This is not `ox-hugo` related because this happend also in the past where I used `ox-publish.el` only. But without reproducable steps it is hard to explain that to the org-roam people.

I am sitting down now, sleeping some nights over this and maybe have to do my own stuff via Python.

**Reply by kaushalmodi** [reply]
I don't know what issue you are seeing for those Org files.

- They are exporting alright for me: https://gitlab.com/hugo-mwe/hugo-mwe/-/commit/c77421dd3993cde95eb52aca93c1cf1da29b6f4c
- Here's the rendered website from that: https://ox-hugo-581--hugo-mwe.netlify.app/posts/20220309102032-foobar.html
   Btw note that those date lines are failing to parse as date front-matter and are showing up verbatim in the content because you are missing the `#` ([ref](https://github.com/kaushalmodi/ox-hugo/issues/587#issuecomment-1066136930)).
   
  ![image](https://user-images.githubusercontent.com/3578197/158186018-3b53d146-86ce-4446-a3a3-80232bb8aa95.png)

**Reply by kaushalmodi** [reply]
My best guess is that the issues you are seeing is due to something in your emacs config, or you are not setting `org-id-extra-files` correctly?

**Reply by kaushalmodi** [reply]
> maybe have to do my own stuff via Python

If you think that Python parsing is easier, then just follow through that effort. It might be beneficial for the Emacs/Org community who find the ox-publish or Hugo approach more difficult. It's good to have a variety of tools to fit the needs of different people 👍

**Reply by buhtz** [reply]
> My best guess is that the issues you are seeing is due to something in your emacs config, or you are not setting `org-id-extra-files` correctly?

Thanks for guessing. I would exclude my emacs config because I also tried a minimal config. Also there sometimes the error occurs and sometimes not.

The point about `org-id-extra-files` is mystic to me. I do not understand that variable always after `C-h v` and what this has to do with my config. My org files are all in the same directory. There are no "extras". I do know have "addiational files" (as stated in the help text about from that variable). And also in the test environment where I tried to figure out the reproducible steps there where only 2 (sometimes 3) org files side by side in the same folder.

I do not know what I should do with that variable. My help system tells me it is defined in `org-id.el` and its value is `org-agenda-text-search-extra-files` (which is `nil`). And `org-id-track-globally` is `t`.

**Reply by kaushalmodi** [reply]
_Note: I don't use org-roam or org-id's, but here's what I understand just by read the variable docstrings (`C-h v`)._

- `C-h v org-id-extra-files` 
   > When Org reparses files to remake the list of files and IDs it is tracking,
it will normally scan the agenda files, the archives related to agenda files,
any files that are listed as ID containing in the current register, and
any Org file currently visited by Emacs.
You can list additional files here.
This variable is only relevant when ‘org-id-track-globally’ is set.

  - So "extras" would be any file that's not part of your Org agenda? The default value of `org-id-extra-files` is `org-agenda-text-search-extra-files`, so that explains.
  - If there's any file you need Org to search for when resolve `id:` links, those files need to be added to `org-id-extra-files`. So in your emacs config, you would need to add all the Org files you plan to link in that list variable.
 

> what this has to do with my config

If you plan to use the Org files for links and you don't set `org-id-extra-files`, that's a user-config error. Again, I don't use org-id based links and don't set that variable. So I could be wrong.

Relevant comment: https://github.com/kaushalmodi/ox-hugo/issues/447#issuecomment-1003689609

**Reply by buhtz** [reply]
In short: I do not need to set it because I have only *one* directory where org files are in.

**Reply by kaushalmodi** [reply]
> In short: I do not need to set it because I have only one directory where org files are in.

But you mentioned that org-id links don't work if you don't set that `org-id-extra-files` list. 
The docstring doesn't say that you don't need to set it if you have only one directory.

**Reply by buhtz** [reply]
> But you mentioned that org-id links don't work if you don't set that `org-id-extra-files` list.
No. I never touched that variable. And as I described in some cases the ID-links are there and sometimes not - no matter if org-id-extra-files is set or not.

And which value should I set if there is no other folder with org-files in it?

> The docstring doesn't say that you don't need to set it if you have only one directory.
That wouuld be a bug. ;)

**Reply by kaushalmodi** [reply]
> as I described in some cases the ID-links are there and sometimes not

Sorry, I couldn't help much here. It might have to do with the Org id locations cache being correct at times when you are interactively adding the Org id links. It will be good to raise this question on a wider forum. I'd assume that that issue is unrelated to ox-hugo. You'd get the exact same behavior with any other exporter too. 

> And which value should I set if there is no other folder with org-files in it?

That's a list variable: a list of all the Org files containing Org IDs. 

> That wouuld be a bug. ;)

By setting the variable correctly all the Org ID linking tests are passing consistently. If you feel it's a bug please provide a reproducible example and post it in the Org mailing list.

**Comment by kaushalmodi** [comment, marked as answer]
If `org-id-extra-files` is set correctly, this issue should not occur. Marking this as solved as this issue cannot be reproduced in the `ox-hugo` test suite (see https://github.com/kaushalmodi/ox-hugo/tree/main/test/site/content-org/org-roam) or locally.
